std/algorithm/sorting.d

```diff
@@ -1048,140 +1048,247 @@ unittest
             (index3));
 }
 
-/**
-Lazily computes the union of two or more ranges $(D rs). The ranges
-are assumed to be sorted by $(D less). Elements in the output are not
-unique; the length of the output is the sum of the lengths of the
-inputs. (The $(D length) member is offered if all ranges also have
-length.) The element types of all ranges must have a common type.
+struct Merge(alias less = "a < b", Rs...)
+    if (Rs.length >= 2 &&
+        allSatisfy!(isInputRange, Rs) &&
+        !is(CommonType!(staticMap!(ElementType, Rs)) == void))
+{
+    public Rs source;
+    private size_t _lastFrontIndex = size_t.max;
+    static if (isBidirectional)
+    {
+        private size_t _lastBackIndex = size_t.max; // `size_t.max` means uninitialized,
+    }
 
-Params:
-    less = Predicate the given ranges are sorted by.
-    rs = The ranges to compute the union for.
+    import std.functional : binaryFun;
+    import std.traits : isCopyable;
+    import std.typetuple : anySatisfy;
 
-Returns:
-    A range containing the union of the given ranges.
- */
-struct Merge(alias less = "a < b", Rs...) if (allSatisfy!(isInputRange, Rs))
-{
-private:
-    Rs _r;
-    alias comp = binaryFun!(less);
-    uint _crt;
+    private alias comp = binaryFun!less;
+    private alias ElementType = CommonType!(staticMap!(.ElementType, Rs));
+    private enum isBidirectional = allSatisfy!(isBidirectionalRange, staticMap!(Unqual, Rs));
+
+    debug private enum canCheckSortedness = isCopyable!ElementType && !hasAliasing!ElementType;
```

Comment:
Guess this would be a version:

```d
debug static if (isCopyable!ElementType && !hasAliasing!ElementType)
    version = can_check_sortedness;
```

Unsure on which is the best way.